blog/2020-11-19-error-message-design.md

```diff
@@ -0,0 +1,62 @@
+
+As framework and library designers, we can and should do better for our users with some careful design principles.
+
+## Why We Should Care
+
+Who among us hasn't thrown up their hands at an unhelpful `Error: NullPointerException` or `undefined is not a function`? Of course, this is a [straw man](https://en.wikipedia.org/wiki/Straw_man) - unanticipated errors always leak internals. We should do our best to handle these errors as much as possible! 
+
+The frustrating part comes from errors that were *anticipated* but obviously not given any thought:
+
+![Example of unhelpful error message from Twitter screenshot](https://dev-to-uploads.s3.amazonaws.com/i/qiivj3cdvkbo07fwrxq6.png)
+
+This is particularly aggravating because it compounds an already unhappy situation with an even worse debugging and bug reporting experience.
+
+Thoughtful error message design is one of the less glamorous, but most important, parts of [Developer Exception Engineering](https://www.swyx.io/developer-exception/). Most framework developers instinctively write error messages that make sense to *them*, but a few minutes of extra effort can save DAYS of debugging for *everyone else* (including maintainers!). 
+
+We must acknowledge two things:
+
+- there is immense information asymmetry between framework developer and framework user (developers know the codebase far better than users)
+- error messages are part of your developer experience too
+
+You might be familiar with the old adage "If you fail to plan, you plan to fail." Here, we have a slight twist: **If you fail to plan to fail, you are planning to fail your users when they need you most.** Touting great developer experience with lousy error message design is like selling a beautiful car with seatbelts that don't work when there's a crash.
+
+## Error Message Style Guide
+
+At Temporal, our current style guide reflects our engineering roots in [Uber](https://github.com/uber-go/guide/blob/master/style.md#error-types) and [Golang](https://blog.golang.org/go1.13-errors) with some tactical recommendations:
+
+- Use `errors.New` for simple messages and `fmt.Errorf` for formatted messages. Make sure `errors` is standard go package.
+- Error messages start with lowercase and no punctuation at the end: `unable to open file` instead of `Unable to open file..`
+- Error messages should be actionable, e.g. `no such directory` instead of `loading service config`.
+- Use `unable to...` instead of `failed to...`, `can't...`, or `could not...`.
+- Messages should reflect logic action that we just tried to make (in a call on previous line which returned an error), not a function name: `load service config` instead of `LoadServiceConfig`. This message should add actionable meaningful context to the existing error message.
```

Comment:
I feel this is the weakest since the wording doesn't make message as clear as it could be.

Reply:
```suggestion
- Messages should reflect the logical action, not merely a function name: `load service config` instead of `LoadServiceConfig`. This message should add actionable, meaningful context to the existing error message.
```